### DISCUSSION TITLE
FulFillmentInbound-GetShipmentItems always throws a parameter exception.

### DISCUSSION THREAD

**Discussion by Ali-DNos** [abuzuhri/Amazon-SP-API-CSharp#249] [Q&A]
Hey, I was migrating from MWS to SP API and found this library to be great and very intuitive to use. 

I have been facing an issue calling FulFillmentInbound.GetShipmentItems. I am adding ShipmentId and MarketPlaceId to the parameters however it throws the following exception:

Either the ShipmentId or a pair of LastUpdatedAfter and LastUpdatedBefore must be passed in input

Just to test I added the LastUpdatedAfter and LastUpdatedBefore parameters in the Parameter class object as well but it still gave the same error. Am I missing something here? Need assistance.

**Comment by abuzuhri** [comment]
can you give us sample for the code used when you get this error

**Reply by Ali-DNos** [reply]
Thank you for the response.  I am adding the necessary functions below:

private static void GetShipmentItems(string _marketPlaceId, string _refreshToken, InboundShipmentInfo _shipment)
{

           AmazonConnection _amazonConnection = ApiConfig.GetAmazonConnection(_marketPlaceId, _authToken);

            ParameterGetShipmentItems _shipmentItemsParameters = new ParameterGetShipmentItems();
            _shipmentItemsParameters.MarketplaceId = _marketPlaceId;
            _shipmentItemsParameters.ShipmentId = _shipment.ShipmentId; //ShipmentId is valid
            //_shipmentItemsParameters.LastUpdatedAfter = DateTime.UtcNow.AddDays(-30);  //tried this as well but same exception
            //_shipmentItemsParameters.LastUpdatedBefore = DateTime.UtcNow; //tried this as well but same exception

            List<InboundShipmentItem> _shipmentItemsList = 
            _amazonConnection.FulFillmentInbound.GetShipmentItems(_shipmentItemsParameters);
           //always throws the exception at the above line.
}

public static AmazonConnection GetAmazonConnection (string _marketPlaceId, string _refreshToken)
        {
            AmazonCredential _amazonCredential = new AmazonCredential()
            {
                AccessKey = _ApiCredential._awsAccessKeyId,
                SecretKey = _ApiCredential._secret,
                RoleArn = _ApiCredential._roleArn,
                ClientId = _ApiCredential._clientId,
                ClientSecret = _ApiCredential._clientSecret,                
                RefreshToken = _refreshToken,
                Environment = Environments.Production,
                MarketPlace = MarketPlace.GetMarketPlaceByID(_marketPlaceId)
            };
            AmazonConnection _amazonConnection = new AmazonConnection(_amazonCredential);

            return _amazonConnection;
        }

**Reply by Ali-DNos** [reply]
I am new to the forums, tried to format the code numerous times, but didn't work. Apologies.

**Comment by smargoli2** [comment]
I am also having this issue. @Ali-DNos , did you resolve this?
```
var paramGetShipmentItems = new ParameterGetShipmentItems
            {
                QueryType = Constants.QueryType.SHIPMENT,
                MarketplaceId = MarketPlace.US.ID,
                ShipmentId = shipment.ShipmentId
            };

```

**Comment by smargoli2** [comment]
I ended up using GetShipmentsAsync, GetShipmentItemsByShipmentIdAsync and then GetShipmentItemsByNextTokenAsync.
```
var paramGetShipments = new ParameterGetShipments
        {
            QueryType = QueryType.DATE_RANGE,
            MarketplaceId = MarketPlace.US.ID,
            ShipmentStatusList = new List<FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.ShipmentStatus> { FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.ShipmentStatus.CLOSED },
            LastUpdatedAfter = startDate,
            LastUpdatedBefore = endDate,
        };
        var result = await _amazonConnection.FulFillmentInbound.GetShipmentsAsync(paramGetShipments);

        var groupedItems = new Dictionary<string, List<InboundShipmentItem>>();

        foreach (var shipment in result.ShipmentData)
        {
            var items = await _amazonConnection.FulFillmentInbound.GetShipmentItemsByShipmentIdAsync(shipment.ShipmentId);
            var nextPageToken = items.NextToken;
            var shipmentItemParams = new ParameterGetShipmentItems()
            {
                MarketplaceId = MarketPlace.US.ID,
                QueryType = QueryType.SHIPMENT
            };

            if (!groupedItems.TryGetValue(shipment.ShipmentId, out List<InboundShipmentItem> value))
            {
                value = new List<InboundShipmentItem>();
                groupedItems[shipment.ShipmentId] = value;
            }

            value.AddRange(items.ItemData);

            while (!string.IsNullOrEmpty(nextPageToken))
            {
                shipmentItemParams.QueryType = QueryType.NEXT_TOKEN;
                var moreItems = await _amazonConnection.FulFillmentInbound.GetShipmentItemsByNextTokenAsync(nextPageToken, shipmentItemParams);
                value.AddRange(moreItems.ItemData);
                nextPageToken = moreItems.NextToken;
            }
        }
```
